client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -20,266 +20,1049 @@
+		local_max_pin_duration: Duration,
+		backend: Arc<BE>,
+	) -> Self {
+		SubscriptionsInner {
+			global_blocks: Default::default(),
+			global_max_pinned_blocks,
+			local_max_pin_duration,
+			subs: Default::default(),
+			backend,
+		}
+	}
+
+	/// Insert a new subscription ID.
+	fn insert_subscription(
+		&mut self,
+		sub_id: String,
+		runtime_updates: bool,
+	) -> Option<oneshot::Receiver<()>> {
+		if let Entry::Vacant(entry) = self.subs.entry(sub_id) {
+			let (tx_stop, rx_stop) = oneshot::channel();
+			let state = SubscriptionState::<Block> {
 				runtime_updates,
 				tx_stop: Some(tx_stop),
-				blocks: HashSet::new(),
-				max_pinned_blocks,
-			})),
+				blocks: Default::default(),
+			};
+			entry.insert(state);
+			Some(rx_stop)
+		} else {
+			None
 		}
 	}
 
-	/// Trigger the stop event for the current subscription.
-	///
-	/// This can happen on internal failure (ie, the pruning deleted the block from memory)
-	/// or if the user exceeded the amount of available pinned blocks.
-	pub fn stop(&self) {
-		let mut inner = self.inner.write();
+	/// Remove the subscription ID with associated pinned blocks.
+	fn remove_subscription(&mut self, sub_id: &str) {
+		let Some(mut sub) = self.subs.remove(sub_id) else {
+			return
+		};
 
-		if let Some(tx_stop) = inner.tx_stop.take() {
-			let _ = tx_stop.send(());
+		// The `Stop` event can be generated only once.
+		sub.stop();
+
+		for (hash, state) in sub.blocks.iter() {
+			if !state.state_machine.was_unpinned() {
+				self.global_unpin_block(*hash);
+			}
 		}
 	}
 
-	/// Pin a new block for the current subscription ID.
+	/// Ensure that a new block could be pinned.
 	///
-	/// Returns whether the value was newly inserted if the block can be pinned.
-	/// Otherwise, returns an error if the maximum number of blocks has been exceeded.
-	pub fn pin_block(&self, hash: Block::Hash) -> Result<bool, SubscriptionManagementError> {
-		let mut inner = self.inner.write();
+	/// If the global number of blocks has been reached this method
+	/// will remove all subscriptions that have blocks older than the
+	/// specified pin duration.
+	///
+	/// If after removing all subscriptions that exceed the pin duration
+	/// there is no space for pinning a new block, then all subscriptions
+	/// are terminated.
+	///
+	/// Returns true if the given subscription is also terminated.
+	fn ensure_block_space(&mut self, request_sub_id: &str) -> bool {
+		if self.global_blocks.len() < self.global_max_pinned_blocks {
+			return false
+		}
 
-		if inner.blocks.len() == inner.max_pinned_blocks {
-			// We have reached the limit. However, the block can be already inserted.
-			if inner.blocks.contains(&hash) {
-				return Ok(false)
-			} else {
+		// Terminate all subscriptions that have blocks older than
+		// the specified pin duration.
+		let now = Instant::now();
+
+		let to_remove: Vec<_> = self
+			.subs
+			.iter_mut()
+			.filter_map(|(sub_id, sub)| {
+				let sub_time = sub.find_oldest_block_timestamp();
```

Comment:
If we have to choose which subscription to remove... shouldn't we have to find the newest block timestamp for a subscription? What I mean is if I have several subscriptions is not the most recent one that should be considered? Otherwise I risk to lower my pinning time for a new request just because an old one I previously made